wire/hybridbus.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2022 - See NOTICE file for copyright holders.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wire
+
+import (
+	"context"
+	"fmt"
+
+	"polycry.pt/poly-go/errors"
+	"polycry.pt/poly-go/sync/atomic"
+)
+
+var _ Bus = (*hybridBus)(nil)
+
+// A Bus that can send messages over a local bus and a remote bus.
+type hybridBus struct {
+	buses []Bus
+}
+
+// NewHybridBus creates a hybrid bus that sends and receives messages over
+// multiple buses. Publishing a message over a hybrid bus will attempt to
+// publish it over all its sub-buses simultaneously. Subscribing a client to the
+// hybrid bus will subscribe it to all its sub-buses. The sub-buses can still be
+// accessed independently for more fine-grained control.
+func NewHybridBus(buses ...Bus) Bus {
+	if len(buses) == 0 {
+		panic("expected at least one sub-bus.")
+	}
+	for i, bus := range buses {
+		if bus == nil {
+			panic(fmt.Sprintf("nil Bus[%d] passed to NewHybridBus", i))
+		}
+	}
+
+	return &hybridBus{buses: buses}
+}
+
+// Publish an envelope over all sub-buses simultaneously.
+func (b *hybridBus) Publish(ctx context.Context, e *Envelope) error {
+	// Whether an attempt to publish the envelope succeeded.
+	var success atomic.Bool
+	// Sub-context for sending, cancel on first success.
+	sending, cancelSending := context.WithCancel(ctx)
+	defer cancelSending()
+
+	errg := errors.NewGatherer()
+	for _, bus := range b.buses {
+		bus := bus
```

Comment:
Maybe add a comment on why this is needed? Otherwise, this looks like a mistake.

Reply:
Not going to comment this, as it is required and common in go. If you call an anonymous function in a goroutine inside a for loop, then all variables are referenced by name, not by value. This means that the for loop iterator variable will change before or during the execution of the goroutine. A local copy is created to pin the value the loop variable had during that iteration.
If we started commenting this, there'd be no end to the comments we would have to make.